src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Helper type for building dynamically-sized arrays while minimizing allocations and copying.
+    /// </summary>
+    internal struct LargeArrayBuilder<T>
+    {
+        private const int StartingCapacity = 4;
+        private const int ResizeLimit = 32;
+
+        private T[] _first;                 // The first buffer we store items in. Resized until ResizeLimit.
+        private ArrayBuilder<T[]> _buffers; // After ResizeLimit * 2, we store previous buffers we've filled out here.
+        private T[] _current;               // Current buffer we're reading into. Is _first <= ResizeLimit.
+        private int _index;                 // Index into the current buffer.
+        private int _count;                 // Count of all of the items in this builder.
+
+        /// <summary>
+        /// Constructs a new builder.
+        /// </summary>
+        /// <param name="initialize">Pass <c>true</c>.</param>
+        public LargeArrayBuilder(bool initialize) : this()
```

Comment:
I just checked. If we initialize members explicitly then the constructor doesn't get inlined:

```asm
G_M36021_IG02:
       488D4DC8             lea      rcx, bword ptr [rbp-38H]
       BA01000000           mov      edx, 1
       E88CFAFFFF           call     LargeArrayBuilder`1:.ctor(bool):this
```

While this isn't the bottleneck, I think using `: this()` looks more readable anyway.